### PR TITLE
ENH CMS 6 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/.github/issues/350

Needs https://github.com/silverstripe/supported-modules/pull/49 merged before CI will work

Changes silverstripe/closure dependency to laravel/serializable-closure

Also a bunch of other maintenance changes to make this compatible with a CMS 6 install to ensure things work correctly in CI

Create a 5.0 branch after this is merged and tag 5.0.0-beta1, though we could arguably just release it straight to stable

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 [![CI](https://github.com/silverstripe/doorman/actions/workflows/ci.yml/badge.svg)](https://github.com/silverstripe/doorman/actions/workflows/ci.yml)
 
-Child process management. Compatible from PHP `5.3` to PHP `7`. Needs no extensions.
+Child process management. Needs no extensions.
 
-> `2.x` supports PHP `5.5.9` and upwards. If you need PHP `5.3/4` support, use a `1.x` release.
+The is currently only intended for use with Silverstripe CMS.
```
Comment (GuySartorelli) [inline README.md]:
```suggestion
This currently only intended for use with Silverstripe CMS.
```
Small typo - but also I don't know what the purpose of that note is. Was this a fork of something and we had to adopt it? Or does it have any silverstripe cms specific code? If no to both of those, it seems like this is fine for general use?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
I assume this needs https://github.com/silverstripe/supported-modules/pull/49 to go green....

Though the error is "No valid PHP version allowed". I thought it used the php dependency from composer.json as a fallback? Seems we should definitely investigate that and fix it if needed rather than arbitrarily dropping this into the supported modules json.

**Comment by emteknetnz** [comment]
> I thought it used the php dependency from composer.json

It would, expect this module does not actually require silverstripe so CI cannot just look at the version of framework in composer.json

File: README.md (hunk 1 above)
Comment (emteknetnz) [reply README.md]: Removed

**Comment by GuySartorelli** [comment]
>> I thought it used the php dependency from composer.json
>
> It would, expect this module does not actually require silverstripe so CI cannot just look at the version of framework in composer.json

What does the version of framework have to do with it? The composer.json file for this repo has `php` as a dependency explicitly in the `require` section.

Edit: I've just had a look and `isAllowedPhpVersion` does check the composer php constraint before anything else.
I'm hesitant to merge https://github.com/silverstripe/supported-modules/pull/49 because it seems like it's just working around a legitimate bug in generate matrix.

**Comment by emteknetnz** [comment]
I think isAllowedPhpVersion() checks to see if a particular version of php is allowed given the current composer.json being used

Before that a list of valid PHP versions is generated, e.g. for cms 6 it is 8.3 and 8.4

However doorman is not aware of what version of the cms is being used, because there isn't a version of framework, or admin etc in composer.json which can be used to derive the version of installer to use

Hence it's given the error "No valid PHP version allowed", cos the list of allowable PHP versions is empty

**Comment by GuySartorelli** [comment]
Ahhhh yup you're right. Probably that's something we ought to change but not right now.
Eventually `getListOfPhpVersionsByBranchName` will swap to using `5` as a fallback (or `MetaData::LOWEST_SUPPORTED_CMS_MAJOR` which will also swap to `5`) and that will use more modern PHP versions.

For now, the change you've made is correct.

**Review by GuySartorelli — APPROVED** [review]
MOG

> Create a 5.0 branch after this is merged and tag 5.0.0-beta1, though we could arguably just release it straight to stable

Wouldn't it be `alpha1` not `beta1`?
In any case I think since this isn't a supported repo it doesn't need to go through the alpha->beta->rc->stable cycle. Feel free to tag as stable if you think that's more appropriate.
We might forget to tag the stable if we don't do it right away so that's probably best.

**Comment by GuySartorelli** [comment]
Matrix still refusing to build

**Comment by emteknetnz** [comment]
I've created a PR which should resolve https://github.com/silverstripe/gha-generate-matrix/pull/115

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**